### ISSUE TITLE
why the performance of segformer-b3  worse than hrnet-18-ocr?

### ISSUE THREAD

**Issue by AncientRemember** [NVlabs/SegFormer#6]
same custom datasets,same optimizer 
why the performance of segformer-b3  worse than hrnet-18-ocr?

**Comment by xieenze** [comment]
I don't think so. Which data set are you referring to?
On ade20K Segformer-b3 get 49.3 mIoU but hrnet-18-ocr only 37.7.
On Cityscapes the conclusion is the same.
So actually I don't know what you are talking about.

**Comment by Chrisding** [comment]
Hi, thank you for being interested. Sorry for the confusion with the previous reply.

Could you let us know more about the details of your custom datasets and implementation? Did you try to optimize both methods well enough? The two methods may have different set of optimal parameters so make sure you optimize each separately. Will do our best to help here.

**Comment by AncientRemember** [comment]
#ocrnet-hr18 config
```
_base_ = [
    '../_base_/models/ocrnet_hr18.py',
    '../_base_/datasets/oda.py', '../_base_/default_runtime.py',
    '../_base_/schedules/schedule_20k_cosperiod.py'
]
# norm_cfg = dict(type='SyncBN', requires_grad=True)
norm_cfg = dict(type='BN', requires_grad=True)
model = dict(decode_head=[
    dict(
        type='FCNHead',
        in_channels=[18, 36, 72, 144],
        channels=sum([18, 36, 72, 144]),
        in_index=(0, 1, 2, 3),
        input_transform='resize_concat',
        kernel_size=1,
        num_convs=1,
        concat_input=False,
        dropout_ratio=-1,
        num_classes=2,
        norm_cfg=norm_cfg,
        sampler=dict(type='OHEMPixelSampler', thresh=0.67, min_kept=35000),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=0.4,
            class_weight=[1, 1.3]
            )
        # loss_decode=dict(
        #     type='LabelSmoothSoftmaxCE', loss_weight=0.4,
        #     class_weight=[1 ,2]
        #     )
    ),
    dict(
        type='OCRHead',
        in_channels=[18, 36, 72, 144],
        in_index=(0, 1, 2, 3),
        input_transform='resize_concat',
        channels=512,
        ocr_channels=256,
        dropout_ratio=-1,
        num_classes=2,
        norm_cfg=norm_cfg,
        sampler=dict(type='OHEMPixelSampler', thresh=0.67, min_kept=35000),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0,
            class_weight=[1 ,1.3]
            )
        # loss_decode=dict(
        #     type='LabelSmoothSoftmaxCE', loss_weight=1.0,
        #     class_weight=[1 ,2]
        #     )
    ),
])
```
#segfor-b2-config
```
_base_ = [
    '../_base_/models/segformer.py',
    '../_base_/datasets/oda.py',
    '../_base_/default_runtime.py',
    '../_base_/schedules/schedule_20k_adamw.py'
]

# model settings
norm_cfg = dict(type='BN', requires_grad=True)
find_unused_parameters = True
model = dict(
    type='EncoderDecoder',
    pretrained='pretrained/mit_b2.pth',
    backbone=dict(
        type='mit_b2',
        style='pytorch'),
    decode_head=dict(
        type='SegFormerHead',
        # type='MLPHead',
        in_channels=[64, 128, 320, 512],
        in_index=[0, 1, 2, 3],
        feature_strides=[4, 8, 16, 32],
        channels=128,
        dropout_ratio=0.1,
        num_classes=2,
        norm_cfg=norm_cfg,
        align_corners=False,
        decoder_params=dict(embed_dim=768),
        loss_decode=dict(type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0,class_weight=[1, 2])
    ),
    # model training and testing settings
    train_cfg=dict(),
    test_cfg=dict(mode='whole'))

# optimizer
optimizer = dict(_delete_=True, type='AdamW', lr=0.00042, betas=(0.9, 0.999), weight_decay=0.01,
                 paramwise_cfg=dict(custom_keys={'pos_block': dict(decay_mult=0.),
                                                 'norm': dict(decay_mult=0.),
                                                 'head': dict(lr_mult=10.)
                                                 }))

lr_config = dict(_delete_=True, policy='poly',
                 warmup='linear',
                 warmup_iters=1500,
                 warmup_ratio=1e-6,
                 power=1.0, min_lr=0.0, by_epoch=False)

data = dict(samples_per_gpu=14)
evaluation = dict(interval=2000, metric='mIoU')


```

**Comment by Chrisding** [comment]
I can see this is a pretty corner case with sample imbalance. Very likely SegFormer is not optimized enough. Different methods behave differently to class imbalance cases. Some may require harder measures to promote the response to smaller classes.

The scratch areas themselves would probably be very small, so resampling 10times may not be enough. You may want to add a reweighting factor to the positive pixels.

The other things you may look into are image resolution and learning rate. Have you played with these factors a bit? Since you are changing it into a binary classification problem, the optimal hyperparameters may change.

**Comment by AncientRemember** [comment]
i use class weight [1,2] at beging of train,and change is to [1,1] after 2000 iter,the result is beter
```
2021-06-16 22:13:30,789 - mmseg - INFO - per class results:
Class                  IoU       pAcc        Acc     Recall
background          100.00     100.00       0.00       0.00
crack                80.51      91.85      98.04     100.00
```

**Comment by AncientRemember** [comment]
> I can see this is a pretty corner case with sample imbalance. Very likely SegFormer is not optimized enough. Different methods behave differently to class imbalance cases. Some may require harder measures to promote the response to smaller classes.
> 
> The scratch areas themselves would probably be very small, so resampling 10times may not be enough. You may want to add a reweighting factor to the positive pixels.
> 
> The other things you may look into are image resolution and learning rate. Have you played with these factors a bit? Since you are changing it into a binary classification problem, the optimal hyperparameters may change.

in my situation,because the scratch data has only 40 images, i put it all into train set and same into the test data
so the eval result looks very high

**Comment by Chrisding** [comment]
I don't think crop size matters in your case. for scratch images my feeling is that they don't require a lot of context.

What I meant by "resolution" is the scale of image you use during training and testing.

**Comment by prnvjb** [comment]
Hey @AncientRemember ,
Which cuda version are you using?

**Comment by Chrisding** [comment]
Closing the issue. No recent follow-up discussions.
